### PR TITLE
finish up flames_drv, pre-testing

### PR DESCRIPTION
**Pull request by pewing**

### CODE REVIEW

File: Flames/flames_drv.py [hunk 1]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
-            
-            #else:
-                #end the poof string with the "end" command.
-                #bangCommandList= + "."
-            
+            controllerDict = defaultdict(list)
+            for controllerId in address: controllerDict[controllerId[:2]].append(controllerId[2])
+
+            for i in controllerDict.keys():
+                onBangCommands.append(
```
Comment (cswales) [inline Flames/flames_drv.py]: Why are you creating two bang command lists? Neither the control protocol nor the controller cares - it's just an event, and the easiest way to handle it is to treat all events the same. (If you're worried about what happens if some simultaneously specifies a poofer to go on *and* go off, dont. Consider that behavior to be an input error, and don't worry about consistency in response to it

File: Flames/flames_drv.py [hunk 2]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
-                #
-            
-            
-            
-            #else:
-                #end the poof string with the "end" command.
-                #bangCommandList= + "."
-            
+            controllerDict = defaultdict(list)
+            for controllerId in address: controllerDict[controllerId[:2]].append(controllerId[2])
```
Comment (cswales) [inline Flames/flames_drv.py]: I personally hate squishing things onto a single line. One gets used to the convention that there will be an indent, and when the reader doesn't see an indent, it creates an unnecessary cognitive load.

File: Flames/flames_drv.py [hunk 3]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
+            for i in controllerDict.keys():
+                onBangCommands.append(
+                    "!" + i + "~".join(map(lambda x: x+"1", controllerDict[i])) + ".")
+                offBangCommands.append(
+                    "!" + i + "~".join(map(lambda x: x+"0", controllerDict[i])) + ".")
 
         except Exception as e:
-            print("Error while sending bang code: ", e)
+            print("Error generating bang code: ", e)
             traceback.print_exc(file=sys.stdout)
```
Comment (cswales) [inline Flames/flames_drv.py]: Use logger.exception() instead of print and traceback

File: Flames/flames_drv.py [hunk 4]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
+                pooferEvent["time"] = startTime
+                pooferEvent["bangCommandList"] = bangCommandList["on"]
+
+                endPooferEvent = {}
+                endPooferEventendPooferEvent["sequence"] = sequenceName
+                endPooferEvent["time"] = endTime
+                endPooferEvent["bangCommandList"] = bangCommandList["off"]
+
+                # TODO: need to figure out best way to sort this thing
+                self.pooferEvents.append(pooferEvent)
```
Comment (cswales) [inline Flames/flames_drv.py:331]: list.sort() is your friend here. It takes a sort parameter, which you can set as the item/attribute getter for the time. Or you can do something more complex.

File: Flames/flames_drv.py [hunk 5]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
+
+    def startFlameEffect(msgObj):
+        if self.checkSequence(msgObj):
+            setUpEvent(msgObj)
+            event_manager.postEvent({"msgType":"sequence_start", "id":msgObj["name"]})
+
+    def stopFlameEffect(msgObj):
+        #  TODO: Need to figure out how to stop this,
+        # since now this is also removing commands to stop poofers
+        event_manager.postEvent({"msgType":"sequence_stop", "id":msgObj["name"]})
```
Comment (cswales) [inline Flames/flames_drv.py]: I'd punt on stopping the poofers that are currently on. They don't stay on for more than 3 seconds anyway, so just removing upcoming poofer events should be fine.

File: Flames/flames_drv.py [hunk 6]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
+        if enabled_poofer != None:
+            event_manager.postEvent({"msgType":"poofer_enabled", "id":msgObj["name"]})
+
+    def resumeAll():
+        self.isStopped = False
+        event_manager.postEvent({"msgType":"global_resume", "id":"all?"})
+
+    def stopAll():
+        self.isStopped = True
+        # TODO: send bang command to stop all poofers
```
Comment (cswales) [inline Flames/flames_drv.py]: I'd consider setting static strings that just blast commands to all the controllers:
!0010~20~30~40~50~60~70~80.
...
!0310~20~30~40~50~60~70~80.

Remember that in the worst case, the poofers will shut off within 3 seconds, so even if you stop no poofers at all, things should be okay.

(In the worst worst case, there's a big red button that cuts power to the solenoids)

(fucking text editor treats my tildes as strike through markers)

File: Flames/flames_drv.py [hunk 7]
```diff
@@ -293,78 +295,125 @@ def checkSequence(self, firingSequence):
...
 
+            for command in bangCommandList:
+                ser.write(command.encode())
 
-    ## parse out firing sequence into bang code ##
-    def generateBangCommandList(self, firingSequence):
+        except Exception as e:
+            ser.close()
+            ser = None
+            print("Error sending bangCommandSequence to poofer controller boards", e)
```
Comment (cswales) [inline Flames/flames_drv.py]: logger.error, not print

File: Flames/flames_drv.py [hunk 8]
```diff
@@ -184,24 +171,39 @@ def run(self):
...
+                logger.exception("Unexpected exception processing command queue!")
+
+
+
     ## sanity and limit checks on an event from firingSequence ##
-    # CSW - suggest just returning True or False, and logging the reason, rather than 
+    # CSW - suggest just returning True or False, and logging the reason, rather than
     # returning the reason as a string. The upper level code really doesn't care why
     # the call failed.
     def checkEvent(self, event):
```
Comment (cswales) [inline Flames/flames_drv.py]: Either get rid of this or rewrite it so the event syntax matches what we're using. It's not critical to check the event structure - this is a very limited application, and one can (mostly) assume that people aren't going to do anything wildly incorrect/stupid. (Hmm. As I'm thinking about that, the one wildly stupid thing that comes to mind is having an event that fires way, way after the other events. You might want to apply some discipline, such as 'no flame sequence can run for over a minute')

File: Flames/flames_drv.py [hunk 9]
```diff
@@ -148,33 +95,73 @@ def shutdown():
...
 class PooferFiringThread(Thread):
     TIMEOUT = 1 # 1 second timeout, even if no events
-    
+
     def __init__(self, cmdQueue):
         Thread.__init__(self)
         logger.info("Init Poofer Firing Thread")
         self.cmdQueue = cmdQueue
         self.running = False
+        self.isStopped = False
```
Comment (cswales) [inline Flames/flames_drv.py]: I'd either rename this variable or add a comment as to why 'isStopped' is not the inverse of 'running'.

**Review by cswales — APPROVED** [review]
Looks good. Changes I'd suggest include:
- Either clean up and use, or remove, the old sequence sanity checking code. Since the event syntax is so different from the current syntax, it's a confusing distraction as is.
- Replace any print() with the appropriate logger.xxx() command
- You really do need to sort your poofer firing list
- Watch the timeout on the main loop. You may end up in situations where the timeout is negative, and I don't know how python handles that. If it's interpreted as 'dont wait', you're fine. If it's interpreted as unsigned, you're fucked.

File: Flames/flames_drv.py (hunk 1 above)
Comment (pewing) [reply Flames/flames_drv.py]: Currently, I'm creating one command to turn the poofer on, and one to turn it off when the duration is over. I'll just change it to let the poofers run their course then?

Comment (cswales) [reply Flames/flames_drv.py]: Hmm. Thats not what I meant. You need a command to turn the poofer off - I was am just confused as to why you have two separate command lists. Seems like there should be one list with both on and off commands in it. Am I missing something?

Comment (pewing) [reply Flames/flames_drv.py]: This function just generates the bang command string. it returns an object that resembles {"on":"bangCommandToTurnOn", "off":"bangCommandToTurnOff"}, which is then used by setUpEvent() to actually append all of the events to pooferEvents.

File: Flames/flames_drv.py (hunk 9 above)
Comment (pewing) [reply Flames/flames_drv.py]: renamed it to isFiringDisabled

**Pull request merged**